### PR TITLE
Fix deferred DRC errors incorrectly classified as SYSTEM_ERROR

### LINKED ISSUE #146: Precheck deferred DRC errors incorrectly classified as SYSTEM_ERROR
## Summary

When the precheck tool runs DRC checks to completion and encounters violations through its "deferred errors" mechanism, the platform incorrectly classifies these as `system` errors instead of `design` errors. This results in:
- `is_manufacturable = NULL` instead of `False`
- `errors = []` (empty) instead of containing the DRC error details
- Users see a confusing "system error" message instead of actionable DRC feedback

## Affected Projects (2-3 December 2025)

| Check ID | Project | User | GDS File | Actual DRC Errors |
|----------|---------|------|----------|-------------------|
| 862 | TinyQV - Crowdsourced Risc-V SoC | MichaelBell | tinyqv_quarter.gds.gz | 18 Magic + 6 KLayout |
| 863 | TinyQV - Crowdsourced Risc-V SoC (0.5x1) | MichaelBell | tinyqv_hwidth.gds.gz | 18 Magic + 6 KLayout |
| 864 | TinyQV - Crowdsourced Risc-V SoC (1x0.5) | MichaelBell | tinyqv_hheight.gds.gz | 18 Magic + 6 KLayout |
| 860 | KianV: A 32-bit RISC-V Linux SoC | splinedrive | chip_top.gds | Magic DRC errors |

## Root Cause

The precheck tool has a "deferred errors" mechanism where DRC checks continue running even after finding violations. At the end, if deferred errors exist, the precheck exits with an exception containing this pattern:

```
Error: The precheck failed with the following exception:
One or more deferred errors were encountered:
18 Magic DRC errors found.
6 KLayout DRC errors found.
```

The platform's log parser classifies this as `type: system` when it should recognize the "deferred errors" pattern and classify it as `type: design`.

### System Error (incorrect classification) - Check ID 862

```
[2025-12-02 09:50:46,598: INFO/ForkPoolWorker-4]   Parsing 974400 bytes of logs...
[2025-12-02 09:50:46,601: INFO/ForkPoolWorker-4]   ✓ Parsing completed: success=False, errors=1, warnings=0
[2025-12-02 09:50:46,607: INFO/ForkPoolWorker-4] Check failed with type: system
[2025-12-02 09:50:46,607: ERROR/ForkPoolWorker-4] System failure detected: Error: The precheck failed with the following exception:
...
[2025-12-02 09:50:47,028: INFO/ForkPoolWorker-4]   Result: FAIL
[2025-12-02 09:50:47,028: INFO/ForkPoolWorker-4]   Errors: 0     <-- BUG: should be non-zero
[2025-12-02 09:50:47,028: INFO/ForkPoolWorker-4]   Warnings: 0
...
[2025-12-02 09:50:47,365: INFO/ForkPoolWorker-4] Task ... succeeded ... {'status': 'completed', 'is_manufacturable': None, 'errors': [], ...}
```

### Design Error (correct classification) - Different project

```
[2025-12-01 12:37:36,204: INFO/ForkPoolWorker-2]   Parsing 12609 bytes of logs...
[2025-12-01 12:37:36,204: INFO/ForkPoolWorker-2]   ✓ Parsing completed: success=False, errors=1, warnings=0
[2025-12-01 12:37:36,209: INFO/ForkPoolWorker-2] Check failed with type: design
[2025-12-01 12:37:36,255: INFO/ForkPoolWorker-2] Design errors found - check completed with errors
...
[2025-12-01 12:37:36,309: INFO/ForkPoolWorker-2]   Result: FAIL
[2025-12-01 12:37:36,309: INFO/ForkPoolWorker-2]   Errors: 1     <-- Correct
[2025-12-01 12:37:36,309: INFO/ForkPoolWorker-2]   Warnings: 0
...
[2025-12-01 12:37:36,676: INFO/ForkPoolWorker-2] Task ... succeeded ... {'status': 'completed', 'is_manufacturable': False, 'errors': [{'message': 'Precheck failed - see full logs for details', 'line': 0, 'category': 'System'}], ...}
```

## Precheck Log Output (for test case)

Full precheck output ending from Check ID 862 (`tinyqv_quarter.gds.gz`):

```
2025-12-02 11:19:03 +0000: Memory Usage (3611076K) :  DRC Total Run time 4444.304003316 seconds
[11:19:06] VERBOSE  Logging subprocess to 'runs/RUN_2025-12-02_09-51-20/11-klayout-drc/xml_drc_report_to_json.log'…                                                                         step.py:1338
───────────────────────────────────────────────────────────────────────────────────── KLayout Design Rule Checker ──────────────────────────────────────────────────────────────────────────────────────
[11:19:06] VERBOSE  Running 'Checker.KLayoutDRC' at 'runs/RUN_2025-12-02_09-51-20/12-checker-klayoutdrc'…                                                                                   step.py:1138
[11:19:06] ERROR    6 KLayout DRC errors found. - deferred                                                                                                                                checker.py:124
─────────────────────────────────────────────────────────────────────────────────────────── Write the layout ───────────────────────────────────────────────────────────────────────────────────────────
[11:19:06] VERBOSE  Running 'KLayout.WriteLayout' at 'runs/RUN_2025-12-02_09-51-20/13-klayout-writelayout'…                                                                                 step.py:1138
PrecheckFlow - Stage 13 - Write the layout ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 13/13 1:27:46
Error: The precheck failed with the following exception:
One or more deferred errors were encountered:
18 Magic DRC errors found.
6 KLayout DRC errors found.


=== SYSTEM ERROR - See error_message field ===
```

## Expected Behavior

When the log contains the pattern:
```
Error: The precheck failed with the following exception:
One or more deferred errors were encountered:
N Magic DRC errors found.
M KLayout DRC errors found.
```

The parser should:
1. Classify as `type: design` (not `system`)
2. Extract the DRC error counts from the log (look for `N Magic DRC errors found` and `M KLayout DRC errors found`)
3. Populate the `errors` field with extracted error details
4. Set `is_manufacturable = False` (not `NULL`)

## Suggested Test Case

Add a test with this log snippet:
```python
LOG_WITH_DEFERRED_DRC_ERRORS = """
[11:19:06] ERROR    6 KLayout DRC errors found. - deferred                                                                                                                                checker.py:124
PrecheckFlow - Stage 13 - Write the layout ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 13/13 1:27:46
Error: The precheck failed with the following exception:
One or more deferred errors were encountered:
18 Magic DRC errors found.
6 KLayout DRC errors found.


=== SYSTEM ERROR - See error_message field ===
"""

def test_deferred_drc_errors_classified_as_design():
    result = parse_precheck_log(LOG_WITH_DEFERRED_DRC_ERRORS)
    assert result.check_type == "design"  # Not "system"
    assert result.is_manufacturable == False  # Not None
    assert len(result.errors) == 2  # Magic + KLayout
    assert any("18 Magic DRC" in e.message for e in result.errors)
    assert any("6 KLayout DRC" in e.message for e in result.errors)
```

## Environment

- Platform: test-platform.wafer.space
- Precheck image: `ghcr.io/wafer-space/gf180mcu-precheck:latest`
- Image digest: `sha256:80a3e206fca8f048933acf317f983527af6111e2f38f2680e532ab95db3bec4e`

### PR DESCRIPTION
**Pull request by mithro**
## Summary

- Fixes precheck deferred DRC errors being incorrectly classified as `system` errors instead of `design` errors
- Adds DRC completion detection: precheck is only considered "complete" if BOTH Magic AND KLayout report results
- If either tool is missing its result → system error (precheck didn't complete, should retry)
- If both tools reported → design error (user can fix DRC violations)

### New DRC Detection Patterns
- `N Magic DRC errors found.` / `N KLayout DRC errors found.` - errors detected
- `Check for Magic DRC errors clear.` / `Check for KLayout DRC errors clear.` - clean
- Metric not found / threshold not set messages - incomplete

### New Functions
- `has_drc_result(logs, tool)` - check if a specific tool reported
- `both_drc_tools_completed(logs)` - verify both tools ran

### Updated Logic
| Scenario | Classification |
|----------|---------------|
| Both tools have results | Design error (if errors) or success (if both clear) |
| Either tool missing | System error (retry) |
| Metric not found | System error (retry) |

## Test plan

- [x] All 958 tests pass locally
- [x] 23 tests in `test_precheck_parser.py` covering all DRC scenarios
- [ ] CI passes

Fixes #146

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Improved DRC detection: consolidated deferred-error reporting that extracts explicit Magic and KLayout error counts and surfaces clear DRC messages.
  * Updated failure classification: missing or incomplete DRC tool results are treated as system errors; design errors now require both tools to have completed.

* **Tests**
  * Added tests covering DRC parsing, deferred-error counts, and classification across success/design/system scenarios.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

File: wafer_space/projects/precheck_parser.py [hunk 1]
```diff
@@ -38,28 +38,124 @@ class ParseResult(TypedDict):
...
+    # Check if both DRC tools completed - if not, it's a system error
+    # Both Magic AND KLayout must report either "errors found" or "clear"
+    if not both_drc_tools_completed(logs):
+        return "system"
 
-    # Default to design error (user must fix)
+    # Both DRC tools reported results - classify based on content
+    # If we got here with exit_code != 0 and both tools completed,
+    # at least one tool has errors → design error
     return "design"
```
[the remark this answers is not in the document]
Comment (mithro) [reply wafer_space/projects/precheck_parser.py]: Fixed: Removed the unused `SYSTEM_ERROR_PATTERNS` constant. The new DRC completion detection logic makes this obsolete - if neither tool reports results, it's a system error; if both report, it's a design error.

File: wafer_space/projects/precheck_parser.py [hunk 2]
```diff
@@ -38,28 +38,124 @@ class ParseResult(TypedDict):
...
+)
+
+# DRC completion patterns - each tool must have ONE of these to be "complete"
+# Matches error counts like "18 Magic DRC errors found."
+MAGIC_DRC_ERRORS_PATTERN = re.compile(
+    r"\d+\s+Magic\s+DRC\s+errors?\s+found",
+    re.IGNORECASE,
+)
+KLAYOUT_DRC_ERRORS_PATTERN = re.compile(
+    r"\d+\s+KLayout\s+DRC\s+errors?\s+found",
```
[the remark this answers is not in the document]
Comment (mithro) [reply wafer_space/projects/precheck_parser.py]: Added a clarifying comment. The precheck tool never outputs '0 errors found' - when there are zero errors, it outputs 'Check for X DRC errors clear' instead. This was verified in the GitHub issue comments which reference the actual precheck source code.

File: wafer_space/projects/precheck_parser.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (mithro) [reply wafer_space/projects/precheck_parser.py]: We intentionally removed `SYSTEM_ERROR_PATTERNS` (addressed in Comment 1) because the DRC completion logic handles all cases:

1. If neither DRC tool reports results → system error (precheck crashed before DRC ran)
2. If both DRC tools report results → design error (precheck completed, user can fix)

The edge case you mention (traceback after both DRC tools report) is practically impossible because DRC checks run near the end of precheck. The test `test_classify_real_system_error_with_traceback` passes because tracebacks occur before DRC results are written.

This simpler logic is easier to reason about and maintain.
